### ISSUE TITLE
[Homework 3] How does ordering of these statements affect seg fault

### ISSUE THREAD

**Issue by choejms** [kaist-cp/cs431#645]
Hi previously I had a seg fault error (invalid memory reference). 
The error was strange because my code passed all the assertions in the `smoke` test but there would be a seg fault error when  drop() function is called, even when my drop() function is blank.

After trying different things, I found that changing the order of following two lines fixes the bug. But I do not understand why it fixes the bug.

Line 1:  `let x = (*raw_pointer).next.lock().unwrap()`
Line 2: `let y = Box::from_raw(raw_pointer)`

I only changed the order of the two lines above and the seg fault error was fixed. 
But I think each line has no effect on the other.
Could you explain how the ordering affects the seg fault?

**Comment by Lee-Janggun** [comment]
Was your change L1;L2 -> L2;L1 or the opposite?

**Comment by choejms** [comment]
It was L1;L2 -> L2;L1

**Comment by Lee-Janggun** [comment]
Now that's very strange... the changed code is actually more incorrect because the one should not access data via the raw pointer after making it a boxed pointer via `Box::from_raw()`.

**Comment by choejms** [comment]
Got it. I will send the code if error persists after passing asan and tsan.
Thank you!

**Comment by gayeun4195** [comment]
Actually, I had same problem with this question and changing the order of code solved the problem. Can I know what is the reason for this weird solution?

**Comment by dubuduru** [comment]
I passed test cases with asan and tsan, and I had exactly the same problem. I found some way to solve the problem of L1;L2:

Line 1: `let x = (*raw_pointer).next.lock().unwrap()`
Line 2: `let y = Box::from_raw(raw_pointer)`
(... some lines of code)
do `drop(x)`

Then the problem does not happen.
If we do not drop x, y is dropped first (as dropping order is reverse of construction order), and it may cause error when we drop x as it has reference of x.. That's what I think. And when we do L2;L1, it drops x first and then y, it won't cause error.
